lib/http_get.js

```diff
@@ -0,0 +1,312 @@
+            /* eslint-enable no-param-reassign */
+            return callClientCb ? callback(err) : null;
+        }
+
+        /* Error persistence OK, return already encountered
+         * error (might be null) */
+        return callClientCb ? callback(currentError) : null;
+    });
+}
+
+function _handleCorruption(reqContext, corruptError, errorAgent) {
+    const toRepair = {
+        rawKey: reqContext.opContext.rawKey,
+        fragments: [[reqContext.chunkId, reqContext.fragmentId]],
+    };
+    _handleErrors(reqContext.opContext, errorAgent, toRepair,
+                  null, corruptError, false);
+}
+
+function getCorruptionCheckedStream(httpReply, reqContext, errorAgent) {
+    const ctypes = protocol.helpers.parseReturnedContentType(
+        httpReply.headers['content-type']);
+    const size = ctypes.get('data');
+    const expectedCRC = ctypes.get('$crc.data');
+
+    if (!size || !expectedCRC) {
+        return httpReply;
+    }
+
+    let bytesUntilCRC = size;
+    let endDataBuffer = null;
+    let crcBuffer = null;
+    const corruptSaferStream = new stream.Transform({
+        transform(chunk, encoding, callback) {
+            if (chunk.length < bytesUntilCRC) {
+                this.push(chunk);
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            endDataBuffer = endDataBuffer === null ?
+                chunk.slice(0, bytesUntilCRC) :
+                Buffer.concat([endDataBuffer, chunk.slice(0, bytesUntilCRC)]);
+
+            crcBuffer = crcBuffer === null ?
+                chunk.slice(bytesUntilCRC) :
+                Buffer.concat([crcBuffer, chunk.slice(bytesUntilCRC)]);
+
+            /* Partial CRC */
+            if (crcBuffer.length < 12) {
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            /* Check actual CRC
+             * First 4 bytes are data crc */
+            const actualCRC = crcBuffer.slice(0, 4).readUInt32LE();
+            if (actualCRC !== expectedCRC) {
+                reqContext.opContext.log.error(
+                    'Corrupted data',
+                    { expectedCRC, actualCRC });
+                const corrupted = new Error('Corrupted data');
+                corrupted.infos = {
+                    status: 422,
+                    method: 'GET',
+                };
+                this.emit('error', corrupted);
+                _handleCorruption(reqContext, corrupted, errorAgent);
+                callback();
+                return;
+            }
+
+            this.push(endDataBuffer);
+            bytesUntilCRC -= chunk.lengh;
+            callback();
+        },
+    });
+
+
+    /* Forward useful fields */
+    corruptSaferStream.statusCode = httpReply.statusCode;
+    corruptSaferStream.headers = httpReply.headers;
+
+    /* Forward error */
+    httpReply.on('error', err => corruptSaferStream.emit('error', err));
```

Comment:
Ok, so that's were we handle the HTTP errors on one of the chunks (other than corruption then, so I guess it'd include short-reads, connection loss eetc?)